rfcs/text/0000-simple-updates.md

```diff
@@ -0,0 +1,159 @@
+
+## Detailed Design
+
+### User Interactions
+
+These are the steps that a wrangler will follow to perform a primary metadata update.
+
+![](../images/0000-simple-updates-ingest-flow.png)
+
+* **Step 1 - Wrangler finds submission**
+
+The wrangler finds the project page in the ingest UI by searching over project names and descriptions. Ingest will provide the search capability.
+
+The project page will list all the submissions for that project (at the moment all our projects have only one submission). They will navigate to this submission page.
+
+* **Step 2 - Wrangler retrieves spreadsheet**
+
+The wrangler will retrieve a spreadsheet for that submission by clicking a “Download metadata” button or similar on the submission page. The spreadsheet they receive will be the original spreadsheet they submitted plus a column at the end containing the UUIDs of all entities. This column exists only to allow ingest to identify how the rows in the spreadsheet relate to existing metadata when the updated spreadsheet is submitted. It should not be edited by wranglers.
+
+
+* **Step 3 - Wrangler edits the spreadsheet with their change**
+
+The wrangler edits the spreadsheet. They can change any metadata cell apart from the ones containing the UUIDs that ingest added. They can add and remove entries from tabs that contain array/module data. For this solution, they must not add or remove whole entities (e.g. biomaterials, protocols) or change the linking between entities. Both these are instances of experimental design representation changes and are out of scope.
+
+* **Step 4 - Wrangler submits spreadsheet**
+
+The wrangler will submit a spreadsheet containing an update through a “Submit update” button or similar on the submission page. For each submission, we will allow only one update to be submitted and processed at a time. For this solution, the wrangler can only submit an updated spreadsheet that validates against the current schemas. 
+
+* **Step 5 - Ingest shows the diff to the wrangler for approval**
+
+The wrangler will be shown the difference between their updates and the existing submission in the ingest UI. They will also be shown any validation errors. If there are no validation errors then they will have the option either to approve the update, in which case ingest will start submitting it to the datastore, or cancel it, in which case the update will be removed from ingest and they can return to step 3. If there are validation errors then the wrangler can only cancel the update and submit a new update spreadsheet.
+
+* **Step 6 - Ingest uploads new versions of metadata files and bundles to the datastore**
+
+For this solution, ingest will calculate the bundle updates and submit these to the datastore. This represents a risk to the scalability of this solution. If performance problems emerge we will need to consider solutions in a future RFC.
```

Comment:
1. Certain metadata changes, such as changes to project metadata, will require updates to all bundles in that project. 
2. Some projects have a large number of bundles (e.g. Tabula Muris at ~100K).
3. We already know from [a previous situation where we had to identify all Tabula Muris bundles for recreation then tombstoneing](https://app.zenhub.com/workspaces/dcp-5ac7bcf9465cb172b77760d9/issues/humancellatlas/ingest-central/343) that this can take a long time, on the order of 2-3 days.
4. So it's possible that metadata updates which touch all or a very large percentage of bundles in a project with many bundles may take us days to complete on the ingest side. At the very least, we have to make 100K of update requests and more 100K requests when we allow analysis updates, assuming copy-forward remains. However, the problem above is different - it was about bundle creation using files already in the datastore, not update) and there may be work we can do ingest side to improve performance.

There was a [ticket](https://github.com/HumanCellAtlas/data-store/issues/1991) open with datastore to investigate an alternative. This is for them to provide an API where ingest could specify a set of file version updates for which all bundles containing the old version should be updated to point to the new version. This would save ingest from having to construct this list out of its own bundle state.

However, this in itself might require significant design and performance assessment. So rather than delay updates further I opted to proceed with doing this task in ingest and seeing what we can achieve in terms of performance. The risk is that we can't achieve good performance or doing so would be prohibitively costly, at which point one option is to revisit the topic of datastore updating bundles if it is much more efficient for them to do so.

Reply:
That sounds good, thanks for clarifying. I think we may need to have a joint discussion about scalability issues here later, including the need to remove the "one bundle per cell" mode of operation in SmartSeq2 as it seems to be detrimental to scalability.